another pkg/ctrl cleanup with implications for pkg/grdchk

mjlosch opened the pull request:
## What changes does this PR introduce?
clean up and new feature

## What is the current behaviour? 
addresses most points in #786
- unnecessary common blocks
- unnecessary repetition 
- identification of ctrl variable in pkg/grdchk

## What is the new behaviour 
- unnecessary common blocks removed
- inquire for `cost final` and call `S/R TURNOFF_MODEL_IO` only once
- new character fields that store `xx_fname` and `whichxyz`, so that index (and `grdchkvarindex`) maps directly to the file and grid type of the control variable to be checked

## Does this PR introduce a breaking change? 
no, just internal

## Other information:
This is a draft so far, to decide, if the method of identifying the control variables via name and grid type is acceptable.
The new field `ncvarfilename` combines `xx_genarr3d_file`, `xx_genarr2d_file`, `xx_gentim2d_file`, and `xx_obcs[n,s,w,e]_file`, and hence is a duplication.
TBD: 
- change the value of `ncvarindex` to either `index` or just `1`. This variable is never used, except to identify if the variable has be activated as control variable. The current convention `ncvarindex = index + 100` causes confusion as to which value to use to access a specific variable.
- print all active control variables to stdout in the control-variable configuration section, currently, the obcs variables are missing.

## Suggested addition to `tag-index`
TBD. (To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

mjlosch commented:
@jm-c please have a look at the use of `ncvarfname` and `ncvartype` in `pkg/grdchk`, before I continue with more "features".

jm-c commented:
I took a look, and I think few more adjustments would simplify things even further:

1. the new variable "nvartype" is only 2 character long and does not allow to distinguish between gentim2d and genarr2d control. I suggest to make it 5 character long: "Arr2D", "Arr3D", "Tim2D", and, for sections, "secXZ" or "secYZ". This way we don't need anymore to set `isgentim2d` and `isgenarr2d` in `ctrl_summary.F`
2. could change `ncvarindex` (and rename it ?) to store the corresponding generic index (if generic control) or just "0" if not (leaving "-1" for un-used). This way we don't need to loop over all generic control names (e.g., in `ctrl_summary.F`) and can directly access the generic control index.
3.  in `ctrl_init_ctrlvar.F`, don't need to check again if  file 'costfinal' exist (now done outside). Would help to have all arguments documented as either Input/Ouput or Updated (I/O/U). And could check that "ivarindex" is in the valid range (i.e., > 0 and <= `maxcvars`).
4.  would be nice to rename `ctrl_init.F` to `ctrl_init_fixed.F` (since there are many "ini" subroutines in `pkg/ctrl`).

jm-c commented:
Forgot one thing: The "hard-coded" correspondence "iarr+100/+200/+300" is also found in `ctrl_pack.F` and in `ctrl_unpack.F`. Is there a way to remove theses ?

mjlosch commented:
> The "hard-coded" correspondence "iarr+100/+200/+300" is also found in `ctrl_pack.F` and in `ctrl_unpack.F`. Is there a way to remove theses ?

Yes, that should be possible. I didn't want to start with this before we agree on how to proceed. The packing/unpacking business is probably the most fragile part of all of this, because it requires to do the same only in `optim_readdata.F` (In my opinion, it would be much safer to do all of the packing/unpacking in a separate program that also does the optimisation step in a more straightforward language, such as python or julia). Once your other suggestions are implemented, we can think about this, but I don't have a reliable test for this.

mjlosch commented:
@jm-c thanks for your input.
> 
> 1. the new variable "nvartype" is only 2 character long and does not allow to distinguish between gentim2d and genarr2d control. I suggest to make it 5 character long: "Arr2D", "Arr3D", "Tim2D", and, for sections, "secXZ" or "secYZ". This way we don't need anymore to set `isgentim2d` and `isgenarr2d` in `ctrl_summary.F`

That's a good suggestion. I would prefer, however, to not use too specific keys like "Arr2D". Imagine that for some reason we need a 2D xy-array that is not generic. Then it would be awkward to us "Arr2D", wouldn't it? I prefer to have 3 characters and then maybe 'xyz', 'xy0' (or just 'xy '), 'xyt', 'sxz', 'sxy' or something like that. But I can be convinced otherwise.

> 2. could change `ncvarindex` (and rename it ?) to store the corresponding generic index (if generic control) or just "0" if not (leaving "-1" for un-used). This way we don't need to loop over all generic control names (e.g., in `ctrl_summary.F`) and can directly access the generic control index.

I also like this idea, but I am not quite sure how it works exactly. What about the obcs-ctrls? Or do you mean just this: `ncvarindex(ivarindex) = ivarindex` (which I prefer). Do you have a suggestion for a new name?

> 3. in `ctrl_init_ctrlvar.F`, don't need to check again if  file 'costfinal' exist (now done outside). Would help to have all arguments documented as either Input/Ouput or Updated (I/O/U). And could check that "ivarindex" is in the valid range (i.e., > 0 and <= `maxcvars`).

Will do that, the check was an oversight, anyway.

> 4. would be nice to rename `ctrl_init.F` to `ctrl_init_fixed.F` (since there are many "ini" subroutines in `pkg/ctrl`).

Will do that.

jm-c commented:
@mjlosch I agree with your latest comment, please go forward with this when you have time.

mjlosch commented:
I have code that implements the idea in @jm-c's latest comment in #786 about specifying the ctrl-variable name (e.g. `xx_theta`) rather than the index. Should I push that, too?

jm-c commented:
@mjlosch I would wait a little bit for that (changing all "data.grdchk"), until we know exactly what to do with hard coded index correspondence in `ctrl_pack.F` and `ctrl_unpack.F`.

mjlosch commented:
With my code, one can specify either name or index, both works. But we can think about the pack/unpack routines first.

jm-c commented:
@mjlosch Regarding the pack/unpack thing, I was thinking that may be we could replace the 3 gen-ctrl loops (" do iarr= 1, ..) with a single loop (like in the summary, "do k = 1, maxcvars") and treat the different cases inside this big loop. What do you think of this ?

mjlosch commented:
Yes, that should work. I was a little bit worried about the order, in which fields are packed or unpacked, but that's probably OK as long as the OBCS-fields come first (compare with `optim/optim_readdata.F`, so far I do not see any general way to get rid of the hardcoded indices for obcs-fields in `optim/optim_readdata.F`).

My main concern about touching the `ctrl_pack/unpack` routines is that our verification suite is not sensitive to changes there, and I don't have a convenient and reliable test for this, either. We could use `tutorial_global_oce_optim`, because I used this as an example for my version of `optim` (https://github.com/mjlosch/optim_m1qn3), but that's in no way general, because `tutorial_global_oce_optim` only uses one (unusual) control variable.
We could also just use cksums of the files that `ctrl_pack` writes. 
Or we could ask @owang01 to test (and debug) whatever we come up with (kidding).

mjlosch commented:
For uniquely identifying the `obcs`-controls in `optim/optim_read/writedata.F`, we could decide to use `ncvarindex(icvar)` together with `ncvargrd(icvar)='m'`. For this we have to "once and for all" declare the `ncvarindex` to be 111-114 for the `obcs`-controls (for backward compatibility with older `ecco_c*` files) and modify `optim/optim_read/writedata.F` accordingly (change 11-14 to 111-114). The should work for `ecco_c*` files produced with current code. Otherwise `ncvarindex` is not used in `optim_read/writedata.F`.

jm-c commented:
@mjlosch I would like to take some time to review these "obcs-ctrl" indices before looking at your latest suggestion (it's like an iterative process here) so please hold on for now before making new changes to this branch.

jm-c commented:
I have few comments:
1. Regarding backward compatibility: given the check for `ncvarindex` value in ctrl_unpack (more precisely, in all `ctrl_set_unpak_*.F`), the changes here are already preventing to "unpack" a file that was generated with current master branch. But in `optim_readdata.F`, I don't see any check for the value of `ncvarindex`, so we
 are fine.
2. I agree it would be good to remove this hard-coded switch for the OBCS control, currently `icvar= 11,12,13,14` in `optim_readdata.F`, and using the value of `ncvarindex` seems a good idea (a little bit like what has been done in this branch for generic control). As I understand, ncvargrd(icvar)='m' already identify OBCS-control (although it's not descriptive of the C-grid location as for other controls).
 However, there is no real need to use number in 111-114 (apart from a narrow backward
 compatibility need). So I would go for changing those to more natural ones, like 1,2,4,3 (for N,S,W,E),
3. Since this would also involve little changes to optim/optim_read/writedata.F we could also write to the "packed" file the value of `ncvartype` vector into the "header" part (there is 1 slot free just after `ncvargrd`). This way the value of
 `ncvarindex` combined with `ncvartype` uniquely identify any particular generic or OBCS control.
4. Regarding optim/optim_readdata.F backward compatibility, to enable to to process current master-branch packed ctrl-files with this branch updated optim/:
 Since ultimately I would like to remove the independent setting of "maxcvars" to just the sum of maxCtrlArr2D + maxCtrlArr3D + maxCtrlTim2D (+4 with OBCS-ctrl), move it to `CTRL_SIZE.h` where it belong to, and get all ctrl-vars params stores continuously in ncvarindex,ncvarrecs,ncvargrd,ncvartype,ncvarfname ... (i.e.r, with no "jump" in all ctrl-vars indices icvar), this will likely involve also little adjustments in optim/optim_readdata.F (and more
 backward compat issue). So at the end, to deal with this backward compatibility issue, could add a CPP option in optim/optim_readdata.F to do that.

BTW, it would be nice to use the same index-variable name when looping over 1:maxcvars. Right now we have: i,k,ivar,icvar (and maybe other names).

NB: the current numbering of OBC-ctrl is not great, N, S, W and E (like jmax, jmin, imin, imax), so if I had to pick a "new" numbering, I would take either (as mentioned above) 1,2,3,4 for N,S,E,W (as in several pkg/obcs src code, but no numbering is involved there) or something that is symetric between N-S and E-W.

mjlosch commented:
Thanks for your comments. I think it all makes sense. I am not sure how to do this backward compatibility. I see these scenarios:

1. If a new optimisation starts from scratch, there's no problem.
2. There is an `ecco-cost/ctrl` file pair (containing gradient information and the previous ctrl vector) after the last run of "master-branch" mitgcmuv_ad. In this case want to be able to read this old gradient/ctrl pair into optim.x (with optim_readdata), but I probably want to write the new control vector  (result of optim.x) into the `ecco_ctrl`-file. 
3. There is an `ecco_cost/ctrl` file pair (containing gradient information and the new ctrl vector) after running "master-branch" optim.x. In this case it would be "cheap" to rerun the new optim.x to produce the new format `ecco-cost` from the old format `ecco-ctrl` file, but we do not have the information to write it properly and we need some CPP-flag/switch in `ctrl_set_unpack_`-routines.

Am I getting this right?

To move forward, I would use your comments to change the system and then we can think about the backward compatibility. What do you think?

jm-c commented:
@mjlosch I've never done anything with `optim/` (only sub-optim) so I don't really get the difference between (2) and (3) but I can see that some info might be missing to write a new "packed-ctrl" file from a current one.
May be there is a quick way to just run only the `optim/optim_readdata.F` (`lheaderonly` = T ?) just to check ?

Anyway, I agree that you could start making further changes (leaving my forth point and backward compat on the side for now).

mjlosch commented:
The latest three commits make both `pkg/ctrl` and `pkg/grdchk` independent of the index, so that from now on, we can change the index and `ncvarindex` in `ctrl_init_fixed.F` as we like without consequences for the results (except for the headers in `ecco_ctrl/cost_MIT_CE_000.opt*`) and the current `data.grdchk` in the verification experiments.

Once we start to change the numbering, e.g. when setting `maxcvars=maxCtrlArr2D + maxCtrlArr3D + maxCtrlTim2D` (+4 with OBCS-ctrl), we will have to adjust all `data.grdchk`. Alternatively, we change all `data.grdchk` to refer to control variable names rather than index as the next step to make `testreport` independent of any ctrl-variable index, then it would be easy to change `maxcvars` etc.

mjlosch commented:
The small differences that I find in the mean/std of adxx in optim have to do with change in order in which the obcs vectors are written. Following @jm-c's suggestion the ordering is now 1,2,3,4=obcsn/s/w/e, before it was 11,12,13,14=obcsn/s/e/w. When I change the numbering back in `ctrl_init_fixed.F` (i.e. 3 = obcsw and 4 = obcse), then the output of optim.x is identical between branches.

The cost function in iteration 1, however, is still different.

jm-c commented:
@mjlosch I am going to take a look, and will let you know soon.

jm-c commented:
I am going to revert some changes related to OBCS-control indices, and if you want to check again if it works better. Note that I did not check myself so I might have missed something.

mjlosch commented:
I will check your version, but here's what's going on: After I changed the numbering so that 3=obcsw and 4=obcse (see above, what's important is the index/ivar order, unfortunately, not the value of `ncvarindex`), optim.x gives the same as the master branch, but in the ensuing iteration 1 is wrong because in `s/r cost_obcsn/s/e/w` (from our beloved `ecco` package) the obcs parameters are referred to as 11-14 (and now should be 1-4 in the correct order). That's a bit annoying, because it breaks the idea of not having to worry about the indices anymore. I'll think of a way to fix this by figuring out the index of the ctrl variable or so.

mjlosch commented:
quick follow up. I will push something that fixes the `ecco/cost_obcs*` issue so that the cost function remains the same even without keeping the original indices. I will keep your changes, although the indices could be changed back to 1-4, but keeping your, i.e. the old, order.
I guess the next change can then be the index, starting from one and incrementing by one, so that only the first places of the maxcvar fields are filled.

mjlosch commented:
@jm-c
> I guess the next change can then be the index, starting from one and incrementing by one, so that only the first places of the maxcvar fields are filled.

I can now do this change (only fill the first indices of the `maxcvars`-fields and it works also with `optim`, i.e. my little test optimisation with `obcs_ctrl` gives the same results), but pushing this requires changing all `data.grdchk`. I could do this to use `grdchkvarname`, so that once and for all, we do not need the `grdchkvarindex` any longer. What do you think?

NB: None of this is backwards compatible any longer and we need to provide something to modify old `ecco_cost*` files.

jm-c commented:
@mjlosch I saw your latest changes. And just to clarify (and correct me if I am wrong): with my changes from today, since I restored the OBCS-ctrl indices to 11-14 (as before) the `ecco/cost_obcs.F` was working again (I think). But since we want to avoid these hard coded indices, we  need your latest changes.

jm-c commented:
And regarding what to do next (we can talk about this next week), it seems to me that it's now time to change all `data.grdchk`
to specify the name of the ctrl-var instead of the "ivar" index value.
But may be I should check now (before the next changes) that all testreport pass (e.g., with Tapenade and in verification_other with TAF, unless you already did).

mjlosch commented:
> And just to clarify (and correct me if I am wrong): with my changes from today, since I restored the OBCS-ctrl indices to 11-14 (as before) the `ecco/cost_obcs.F` was working again (I think). But since we want to avoid these hard coded indices, we need your latest changes.

Yes, with your commit [48169c0](https://github.com/MITgcm/MITgcm/pull/796/commits/48169c03a1c6a2a82eebf48f04907500ba6ca77c) everything worked again, because (a) the order of storing (and hence computing the aggregated numbers like mean, std in optim.x) and (b) the reference in `cost_obcs.F` was restored. 
With my latest commit [48169c0](https://github.com/MITgcm/MITgcm/pull/796/commits/48169c03a1c6a2a82eebf48f04907500ba6ca77c), we are now independent of the indices, so that the index itself does not matter. But unfortunately the order does matter (hence the index) a little bit at the truncation level, so that the order is still important to get the same aggregated numbers, and hence results at the truncation level.

With this we can now completely change the indexing (requires `data.grdchk` to change), and also define `maxcvars` in `CTRL_SIZE.h` as a function of number of obcs-ctrls and number of `genArr`-ctrls. The latter is a big change, because it requires putting `CTRL_SIZE.h` into order 50-100 files. I suggest to do this in two steps.

mjlosch commented:
> And regarding what to do next (we can talk about this next week), it seems to me that it's now time to change all `data.grdchk` to specify the name of the ctrl-var instead of the "ivar" index value. But may be I should check now (before the next changes) that all testreport pass (e.g., with Tapenade and in verification_other with TAF, unless you already did).

Now there's some overlap in our messages. I didn't check tapenade or verification_other. Please go ahead and do so. I'll wait, the weekend just started anyway ...

mjlosch commented:
@jm-c I have done the necessary changes to all `data.grdchk` (but only tested the TAF versions) and am waiting for your OK to push them (also `verification_other`)

jm-c commented:
I've run both TAF and Tapenade adm and tlm tests, all are good and also `verification_other` TAF adm tests good too. So please go forward with `data.grdchk` update.

jm-c commented:
@mjlosch as we talked ealier, going to make little adjustment in few src files. 

The main one will be related to OBCS-ctrl, where I would like to replace test on `ncvarfname` = `xx_obcs{N,S,E,W}_file` with a test on value of `ncvarindex` (=1,2,3,4), combined with ncvargrd='m' (a little bit like in `ctrl_init_wet.F`).
I think this could help to solve the backward compatibility issue since both `ncvargrd` and `ncvarindex` are written in ctrl-pack header file (even in current master branch) whereas `ncvarfname` is not. Of course the value of  `ncvarindex` will not match but would be easy to just do a conversion on index-value (like a little function).

mjlosch commented:
Do we know why the generic controls need no `xx_genarr2_weight` file to work for OpenAD in `hs94.x1x64x5`? That's a bit weird because the control variable `xx_genarr2d_01` should not be initialised in `ctrl_init_fixed.F`, if `xx_genarr2_weight(iArr)=' '` (and this seems to be the case).

jm-c commented:
@mjlosch I really don't know why, with OpenAD, the check for the weight file is skipped (within "#ifndef ALLOW_OPENAD" block) in `ctrl_init.F` (here `ctrl_init_fixed.F`), but also the control file/name are initialised with a "default" string (instead of just the empty string) in `ctrl_readparms.F`. An other OpenAD mystery ?

mjlosch commented:
Independent of OpenAD: I never liked the check for the weight-file. Why don't we check for the `xx_genXXXXX_file` instead. But that's maybe somethings for a different PR.

jm-c commented:
@mjlosch I would like to start trying something for the backward compatibility issue, and the easier would be with a new CPP option (affecting some of the array dimensions). 
For the new CPP name, I could think of "READ_OLD_CTRL_PACK" or "READ_OLD_CTRL_PACK_FILE" or maybe something else you prefer ? And we could define (somewhere) "old_maxcvars = 400" for reading the (old) ctrl-packed header file.

mjlosch commented:
I like `READ_OLD_CTRL_PACK_FILE` better. I don't know your strategy, but for `old_maxcvars`, wouldn't it be better to redefine `maxcvars` if `READ_OLD_CTRL_PACK_FILE` is defined, as it is needed to define many 1D-fields. This redefinition should go to where `maxcvars` is defined, so currently `CTRL.h`, and very soon `CTRL_SIZE.h` (I am working on that, but won't finish before tomorrow). Or: if you only want to do this in `optim_readdata.F`, then a local definition of `old_maxcvars` and the ensuing `ncvarindex`, etc fields could also be done locally to read the old header.

jm-c commented:
Ok, I will start to play with `READ_OLD_CTRL_PACK_FILE`, but I what to keep `maxcvars` separatetely for "old_maxcvars" (just used to read-in header of old-file).

mjlosch commented:
Update: I could push moving `maxcvars` to `CTRL_SIZE.h`. This involves changes to 80 files (also some customised `CTRL_SIZE.h` in `verifcation`). Not sure if this is a good idea now, or rather change the location together with the value of `maxcvars`. In the context of your upcoming backward compatibility, what do you this @jm-c?

jm-c commented:
@mjlosch  regarding  maxcvars isetting and  `CTRL_SIZE.h`, we could also leave `maxcvars` setting inside `CRTL.h` and set it as a function ot the 3 gen-ctrl size from `CTRL_SIZE.h`. This way users will not have to changed they customized version of ``CTRL_SIZE.h`. It's a little bit less safe, but not very clear to me how much (because even if `maxcvars` moved to `CTRL_SIZE.h` but one forgot to update the customized ``CTRL_SIZE.h`, it will trigger also a compiler error).

mjlosch commented:
Don't get me wrong, I think it's a very good idea to move the definition of `maxcvars` to `CTRL_SIZE.h`. When `maxcvars` becomes a function of the 3 gen-ctrl sizes, we will need to include `CTRL_SIZE.h` with every include of `CTRL.h` anyway. Also the ifdefs in `CTRL_SIZE.h` should be removed anyway, shouldn't they.
The question is only, when we want to make this change:
1. now, before you start the backward compatibility work. But keep the definition of `maxvars=400` for now.
2. now, before you start the backward compatibility work. But also make `maxcvars` a function of the three gen-ctrl sizes. (not sure if this is an advantage: the changes to the customised `CTRL_SIZE.h` done in one go)
3. wait until the backward compatibility is done.

My preference is (1) or (2).

jm-c commented:
@mjlosch  I was in favor of (3), not change `CTRL_SIZE.h` from any of the `verification/*/code*` dir so that this PR would not force  users to updated their local `CTRL_SIZE.h`.
Now if this does not work (i.e., the local version in verification exp. needs to be updated anyway) then I would be in favor of 2.

jm-c commented:
And to clarify further: we still need to update all src code to include `CTRL_SIZE.h` anytime `CTRL.h` is included. And if this works without any update of local `CTRL_SIZE.h`, we could remove the CPP condition with `ALLOW_GEN[ARR2/ARR3D/TIM2D]_CONTROL` in all local `CTRL_SIZE.h` in a future PR.

mjlosch commented:
OK, I we do not want to change `CTRL_SIZE.h` in the PR (fine with me), then the definition of `maxcvars` needs to remain in `CTRL.h` for now. I can push the other change (including `CTRL_SIZE.h` wherever there's a `CTRL.h` included), and then we can continue from there, i.e. change the definition of `maxcvars` etc. in harmony with the backward compatibility. Do you agree?

jm-c commented:
Yes, sounds good.

mjlosch commented:
@jm-c, just to be clear: with the last commit [7c2426d](https://github.com/MITgcm/MITgcm/pull/796/commits/7c2426dda08d3c8e60d6374ee6a73f1376251d7f) I went beyond what we agreed on. I did this to show that it works to reduce `maxcvars` (and see if it works with OpenAD). If you have a better way of doing this, I have no problem with reverting [7c2426d](https://github.com/MITgcm/MITgcm/pull/796/commits/7c2426dda08d3c8e60d6374ee6a73f1376251d7f). Without the two commits before that ([2d1be62](https://github.com/MITgcm/MITgcm/pull/796/commits/2d1be620b1b7c2d4328d9d054946f3ba5508b03b) and [d8b707c](https://github.com/MITgcm/MITgcm/pull/796/commits/d8b707cf8b5f73fcb89f7e73bd255031ac4da67b)) the last one does not work.

jm-c commented:
@mjlosch This is good. I currently looking at the changes you pushed yesterday and will let you know if I see something that I would like to do differently.

jm-c commented:
@mjlosch  I am going to push some little changes + a template for backward compat, but will do that tomorrow since it's already late.

jm-c commented:
@mjlosch I made some simplifications (but did not check yet Tapenade or OpenAD) and, in last commit, I changed the 23-line long definition of `maxcvars` to only keep 1 CPP condition on `ALLOW_OBCS_CONTROL` (now only 6 line long), not so much to "save" memory but more to remind that this "4" is used for OBCS-ctrl.
Please take a look and If you don't like this new `maxvars` setting, we can revert this last commit.

And after that I will push some changes to `ctrl_pack/unpack.F`

Side note: I would like to remove all code within #ifdef ALLOW_CTRL_WETV  and also ALLOW_EFLUXP0_CONTROL
(I have been following MITgcm code for some time and have never seen "hFacV" array defined anywere, so no chance line 126 in `ctrl_init_wet.F` would compile with any code version from the last 20+ years). So will make an other set of changes in ` ctrl_init_wet.F`.

jm-c commented:
@mjlosch I went ahead with the other set of changes. I "tested" the backward compat feature by adding this extra CPP option `EXCLUDE_LSOPT_TO_CHECK` in optim (turned on when using "prep_make [Bld-dir] -fake") to just read-in inputs and skipping all the `lsopt` bits. Please take a look and see if it works with `optim_m1dn3`.

Otherwise, the names `prep_make` and `makefile_templ` could be changed, but I like that, in `optim`, the compiler options and other setting to use in the Makefile are taken from the MITgcm Makefile form where all  the includes header files are taken.

mjlosch commented:
The `prep_make` and `makefile_templ` combination does not work for me. User error? The macros `_GET_CPPCMD`, `_GET_FC` etc. do not work for me on my Apple PowerBook. We will also need something for the suffixes. (`.for` instead of `.f`)

jm-c commented:
@mjlosch could you send me (email is fine) the Makefile that `genmake2` produces in, e.g., `tutorial_global_oce_optim` ? I could take a look and try to have something more robust.

mjlosch commented:
I just sent a makefile by email. 

However, I got it to work. The different sed is the main problem, but the suffix (case insensitive file system) is another problem. I could suggest a `makefile_templ/prep_make` that may work for both systems (currently I have no linux system easily available) and we could iterate a bit.

jm-c commented:
@mjlosch OK, I got your Mac-OS Makefile.

> However, I got it to work

Could you push your working version ? We could always tweak these 2 files afterward. And if it's too tricky we could have a second template (for Mac-OS) with a new option in `prep_make` to select which template to use.

mjlosch commented:
with [71c561a](https://github.com/MITgcm/MITgcm/pull/796/commits/71c561a7f2d7203b8d37d363241e9082c60ba0ad), the only thing I need to change in `makefile_templ` is `SUFF=for`, if my file system is case insensitive. If you can think of a good way of getting `SUFF` from the MITgcm-Makefile, it would be great, but not necessary.

mjlosch commented:
With the current code, I can construct an `optim_m1qn3` version that reproduces old results, and I can also read an "old" `ecco_cost_MIT_CE_000` file and write a new `ecco_ctrl_MIT_CE_000` file (once `READ_OLD_CTRL_PACK_FILE` is defined). So this part looks good. I think that this should be enough backward compatibility.

jm-c commented:
@mjlosch few comments:
1. the changes you made in `prep_make` and `makefile_templ` are good, and work for me. I will try tomorrow to see how to get the "SUFF" set from MITgcm `Makefile`.
2. It's great that you were able to try the `READ_OLD_CTRL_PACK_FILE` with `optim_m1qn3` and that it seems to work. I am sure this could be further improved (with more/better printed info) but as this backward-compat feature might not be used too much I agree that it is OK as it (provided it works).

mjlosch commented:
Here's a simple but not very robust (in case we do not want to use `.for` as a suffix) solution for `prep_make`:
```
#-- Get suffix
sstr=`grep "SUFFIXES" $mkFile | grep .for`
sfx=f
[ ! -z "$sstr" ] && sfx=for
sed "s|_GET_SUFF|${sfx}|" $newMkF > $tmpFil
mv -f $tmpFil $newMkF
```
And then we use `_GET_SUFF` in `makefile_templ`

jm-c commented:
@mjlosch I changed (simplify ?) the template to just keep one set of rules. Also tried to get the suffix from the MITgcm `Makefile`. It seems to work for me (I also tried with the Mac-OS Makefile you sent me) so if you could check that it works for you.
And few side comments:
1. If this  works, it would also be compatible with other suffix (whereas your version was only working with ".f" and ".for").
2. I changed the variable name from "SUFF" to "FS" to be the same as in `genmake2` but I am not sure I am happy with that.
3. the way to get the suffix from Makefile should be safe as long as it was generated with `genmake2` and we don't change too much the block of lines 3347-3383 in `genmake2`

mjlosch commented:
Works for me and certainly better than my simple hack. I spent some time unsuccessfully trying to condense line 88 (o:
I am fine with `FS` in the Makefile, although this does not have the same meaning as `FS` in `genmake2` (shell variable vs. make-macro/variable). So maybe use `SUFF` instead of `FS` in the Makefile file and `FS` in `prep_make`? Not important.

Side note:
I am not interested in doing this, but since we are dealing this with anyway: Does it make sense to do something similar to `lsopt`? Currently I can use `make -f Makefile_macos` to generate the library `../lsopt/liblsopt_ecco.a`, but the link step in `optim` still does not work:
```
ld: in ../lsopt//liblsopt_ecco.a(lsopt_top.for), archive member 'lsopt_top.for' with length 20904 is not mach-o or llvm bitcode file '../lsopt//liblsopt_ecco.a' for architecture arm64
```
Since I do not plan to ever use `lsopt`, I do not care and will not investigate, but for the greater good it might make sense to combine `optim` and `lsopt` into one consistent make process. This was actually one of the reasons why I started `optim_m1qn3`.

mjlosch commented:
Not really a concern, but just a report. Because by default (`pkg/ctrl/CTRL_SIZE.h`) `maxCtrlArr2D + maxCtrlArr3D + maxCtrlTim2D = 3`, `optim_readdata` now prints for setups that use the default like `verification/obcs_ctrl`:

```
    OPTIM_READDATA: Reading cost function and gradient of cost function
                  for optimization cycle:            0
                  ... header only
  
  OPTIM_READDATA: opened file ecco_cost_MIT_CE_000.opt0000                                                                                                    
  OPTIM_READDATA: nvartype            5
  OPTIM_READDATA: nvarlength        40960
  OPTIM_READDATA: yctrlid MIT_CE_000
  OPTIM_READDATA: filenopt            0
  OPTIM_READDATA: fileff    1.4770107378617978     
  OPTIM_READDATA: fileiG            1
  OPTIM_READDATA: filejG            1
  OPTIM_READDATA: filensx            2
  OPTIM_READDATA: filensy            2
  [...]
  OPTIM_READDATA: ncvarindex            1           2           4           3           1           0           0
  OPTIM_READDATA: ncvarrecs             4           4           4           4           1           0           0
  OPTIM_READDATA: ncvarxmax            32          32           1           1          32           0           0
  OPTIM_READDATA: ncvarymax             1           1          32          32          32           0           0
  OPTIM_READDATA: ncvarnrmax            8           8           8           8           8           0           0
  OPTIM_READDATA: ncvargrd   m,m,m,m,c,^@,^@,
  OPTIM_READDATA: ncvartype  SecXZ,SecXZ,SecYZ,SecYZ,Arr3D,^@^@^@^@^@,^@^@^@^@^@,
  OPTIM_READDATA: end of optim_readdata
```
That is, we have 5 control variables (4 obcs and one genarr3d), but `maxcvars=7` and we print funny (because uninitialised) values for `ncvargrd` and `ncvartype`. For the other variables, we have zeros. I don't think that this is a problem and would be easily solved by have a customised `CTRL_SIZE.h` with  `maxCtrlArr2D = 0`, and `maxCtrlTim2D = 0`.

jm-c commented:
@mjlosch Thanks for trying this and thanks for the report.

1. I don't have much interest in updating `lsopt`, and since you are providing a better `optim_m1qn3`, I would stay away from any update there. If someone wants to do it it's fine, but I don't really see the need for us.
2. Regarding the funny prints when running `optim.x`, I have a different view: there is some cleaning that could be done in `optim/` code and this would/could involve (from my point of view) initialising local variable. I prefer not to change `CTRL_SIZE.h` to avoid fixing an uninitialised var problem. Note that `pkg/ctrl` (`ctrl_unpack.F`) does not have this problem.
3. I am going to rename (again) "FS" to something else, and following your  idea, will use different name inside `prep_make` and in the final `Makefile`. What about "SUFX" (even easier to grep for) there instead of "SUFF" ?

mjlosch commented:
@jm-c good, 
I agree, let's leave `lsopt` alone. 
We can think about initialising these variables (defined in `CTRL.h`) before `optim_readdata` is called, e.g. in `optim_sub.F`
and `SUFX` sound good to me.

jm-c commented:
@mjlosch I remaned "FS" and
> We can think about initialising these variables (defined in CTRL.h) before optim_readdata is called, e.g. in optim_sub.F

I am not 100 % sure that this will be enough, since the "funny" printed lines are coming from the first call to `optim_readdata` with `lheaderonly=T`, from `optim_numbmod.F`. So may be having this initialisation at the top of `optim_numbmod.F` ?

jm-c commented:
And could you try it and if this works, push it to this branch ?

mjlosch commented:
yes, the initialisation needs to be before the first call of `optim_readdata.F`, so in `optim_numbmod.F` (or in `optim_readparms.F` in optim_m1qn3, where this works). I'll push something.

jm-c commented:
I tried with latest changes and the "funny" prints are gone. 
@mjlosch may be it's time to remove the "Draft" status and see what needs to be done (e.g., some doc) to finish this PR ?
I think the README.md in `optim/` could be updated ; the 3 `data.*` files there (ctrl, optim and ecco) could be removed and we will need a new PR in `verification_other` to update some `data.grdchk`.

mjlosch commented:
I have created PR https://github.com/MITgcm/verification_other/pull/67

owang01 commented on optim/README.md:
```diff
@@ -1,54 +1,76 @@
+With PR #[796](https://github.com/MITgcm/MITgcm/pull/796), the header format of
+the control and gradient vector files and the length of the stored fields has
+changed. To read old gradient vector files (typically called
+`ecco_cost_ecco_cost_MIT_CE_000.opt0000` for the first optimisation call) with
```
`ecco_cost_ecco_cost_MIT_CE_000.opt0000` should be `ecco_cost_MIT_CE_000.opt0000`.

jm-c commented:
@mjlosch I added few words to the `README.md` and adjusted some makefile/`Makefile` in it. If you don't like some of these changes, feel free to go back.

jm-c reviewed (approved):
LGTM.
I've just run testreport for TL and AD tests and all good and the `verification_other` AD tests are good too.

owang01 commented on optim/optim_readdata.F:
```diff
@@ -156,40 +164,68 @@ subroutine optim_readdata(
 #endif
 #ifdef ALLOW_OBCSW_CONTROL
-      read( funit ) ((nWetobcswGlo(k,iobcs), k=1,nr),iobcs= 1,nobcs)
+      read( funit ) ((nWetobcswGlo(k,iobcs), k=1,Nr),iobcs= 1,nobcs)
 #endif
 #ifdef ALLOW_OBCSE_CONTROL
-      read( funit ) ((nWetobcseGlo(k,iobcs), k=1,nr),iobcs= 1,nobcs)
+      read( funit ) ((nWetobcseGlo(k,iobcs), k=1,Nr),iobcs= 1,nobcs)
 #endif
```
It might be better to put OBCSE before OBCSW to be consistent with the new OBCS order 1/2/3/4 for N/S/E/W. Same for the S/R optim_writedata.F.

mjlosch replied:
The same backward compatibility argument holds as before, I think.

owang01 replied:
That's right.

jm-c commented on pkg/ctrl/ctrl_init_fixed.F:
```diff
@@ -0,0 +1,388 @@
+
+C----------------------------------------------------------------------
+
+#ifdef ALLOW_OBCSW_CONTROL
+C--   Western obc.
+      CALL ctrl_init_rec ( xx_obcsw_file,
+     I     xx_obcswstartdate1, xx_obcswstartdate2, xx_obcswperiod, 4,
+     O     xx_obcswstartdate, diffrec, startrec, endrec,
+     I     myThid )
+      ivar = ivar+1
+      iarr = 4
+      CALL ctrl_init_ctrlvar (
+     I     xx_obcsw_file, ivar, iarr, diffrec, startrec, endrec,
+     I     1, sNy, Nr, 'm', 'SecYZ', costfinal_exists, myThid )
+#endif  /* ALLOW_OBCSW_CONTROL */
+
+C----------------------------------------------------------------------
+
+#ifdef ALLOW_OBCSE_CONTROL
+C--   Eastern obc.
+      CALL ctrl_init_rec ( xx_obcse_file,
+     I     xx_obcsestartdate1, xx_obcsestartdate2, xx_obcseperiod, 4,
+     O     xx_obcsestartdate, diffrec, startrec, endrec,
+     I     myThid )
+      ivar = ivar+1
+      iarr = 3
+      CALL ctrl_init_ctrlvar (
+     I     xx_obcse_file, ivar, iarr, diffrec, startrec, endrec,
+     I     1, sNy, Nr, 'm', 'SecYZ', costfinal_exists, myThid )
+#endif /* ALLOW_OBCSE_CONTROL */
+
+C----------------------------------------------------------------------
```
@owang01 I think @mjlosch tried first what you described. It caused 2 issues: 1) machine truncation differences
because of some ordering of terms in addition ; and 2) the ctrl-packed files beeing differents (not just the header info) so this would likely requires more work to get some kind of backward compatibility option to work.

owang01 replied:
@jm-c Thanks. Then it makes sense to leave the OBCSE block after the OBCSW block. Sorry for missing Martin's comment.

mjlosch commented on pkg/ctrl/ctrl_init_ctrlvar.F:
```diff
@@ -1,147 +1,174 @@
+        CALL PRINT_ERROR( msgBuf, myThid )
+        WRITE(msgBuf,'(2A,I4)')
+     &       'S/R CTRL_INIT_CTRLVAR: allowed values: ',
+     &       '0 < ivar <= maxcvars = ', maxcvars
```
good point, but I am not sure, which is better? Maybe we should change the if-statement to `ivar.LE.0` as a valid `ivar` should always be greater than zero, right?

jm-c replied:
I think it's better not to change neither the test ( IF ivar ...) nor the error message, because ivar=0 has a special meaning here, see  line 46.
got changed with this commit: ef029d39f7a89999bb080dc4d7174f69689b6702

mjlosch commented on pkg/ctrl/ctrl_unpack.F:
```diff
@@ -165,88 +183,89 @@ subroutine ctrl_unpack( first, mythid )
 #endif
 #ifdef ALLOW_OBCSW_CONTROL
-          read(cunit) ((filenWetobcswGlo(k,iobcs),
+          read(cunit) ((file_nWetobcswGlo(k,iobcs),
      &         k=1,Nr), iobcs= 1,nobcs)
 #endif
 #ifdef ALLOW_OBCSE_CONTROL
-          read(cunit) ((filenWetobcseGlo(k,iobcs),
+          read(cunit) ((file_nWetobcseGlo(k,iobcs),
      &         k=1,Nr), iobcs= 1,nobcs)
 #endif
```
see above, we need this for easier backward compatibility.

jm-c commented:
Following today discussion, in  `optim/`, I will add few notes to the `README.md` to summarize what `prep_make` does with `makefile_templ`.

owang01 commented:
@mjlosch I tested this PR with `ecco_ctrl` and `ecco_cost` from an ECCO production run, and the results are good. I used the old-formatted `ecco_ctrl` and `ecco_cost` files as the input for the optimization and was able to generate the new-formatted `ecco_ctrl` for the next iteration. The model then successfully unpacked the new-formatted `ecco_ctrl` and produced `xx` files identical to those I obtained using the master branch with the old-formatted `ecco_ctrl`.

jm-c commented:
Thanks @owang01 for checking this. Also I tweak a little bit the README.md  but @mjlosch you can change things that you don't like.
One last thing: I might add one or 2 lines of comments in `CTRL.h` to clarify few things. Otherwise, I think this PR is on a good track for merging, unless I missed something.

mjlosch commented:
This looks good to me now (just fixed some small MarkDown issues in the `README.md` and some more tweaks). I also think that this can be merged now.

jm-c commented:
@mjlosch I made my changes (comments) to CTRL.h and could merge this PR soon after your approval. Also, do you want to change something to the "tag-index suggested addition" ? (will change "grid type" to "shape type" to describe `ncvartype`).

mjlosch commented:
I added a little bit of documentation to CTRL.h. 
Otherwise, this looks good to me I have nothing to add.
I changed grid type to shape type for the tag-index text.
